### PR TITLE
Add verbosity to interface thickness in Cahn-Hilliard (and CHNS) solver

### PR DESCRIPTION
**Pull request by PierreLaurentinCS**
### Description

It is often useful for the user to know the exact value of the interface thickness when it is determined automatically. Hence a verbosity parameter was added to the interface thickness parameter to display it at every iteration and make sure it remains coherent with the problem under study.
### Testing

An application test was added to make sure any future change to the code do not break the interface thickness computation (whether in deal.ii or lethe)

### Documentation
This pull requests adds a new parameter in the Cahn-Hilliard subsection of the .prm file. The documentation was modified accordingly and also made clearer.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Lethe documentation is up to date
- [x] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [x] The branch is rebased onto master
- [x] Changelog (CHANGELOG.md) is up to date
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [ ] The PR description is cleaned and ready for merge

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,11 +3,12 @@
 All notable changes to the Lethe project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/).
 
-## [Master] - 2024-09-24
+## [Master] - 2024-09-25
 
 ### Added
 
-- MINOR The outlet boundary condition was implemented in the matrix-free operators. A test in 2D with a cylinder close to the outlet was added to test this boundary condition for both the matrix-based and the matrix-free application. [#1287](https://github.com/chaos-polymtl/lethe/pull/1287)
```
Comment (AmishgaAlphonius) [inline CHANGELOG.md:10]: I think you've accidentally deleted a previous log entry ^^'

File: doc/source/parameters/cfd/cahn_hilliard.rst [hunk 2]
```diff
@@ -38,14 +38,19 @@ The default values of the Cahn-Hilliard parameters are given in the text box bel
     set potential smoothing coefficient = 1
 
     subsection epsilon
-      set method = automatic
-      set value  = 1
+      set verbosity = verbose
```
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/cahn_hilliard.rst:41]:
```suggestion
      set verbosity = quiet
```

File: applications_tests/lethe-fluid/cahn_hilliard_epsilon_verbosity.prm [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  set fluid dynamics = false
+end
+
+#---------------------------------------------------
+# Cahn Hilliard
+#---------------------------------------------------
+
+subsection cahn hilliard
+  set potential smoothing coefficient = 0.0
+
```
Comment (AmishgaAlphonius) [inline applications_tests/lethe-fluid/cahn_hilliard_epsilon_verbosity.prm:30]:
```suggestion
```

File: applications_tests/lethe-fluid/cahn_hilliard_epsilon_verbosity.prm [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    set density             = 1
+    set kinematic viscosity = 1
+  end
+  set number of material interactions = 1 #by default it is set to 0
+  subsection material interaction 0
+    set type = fluid-fluid
+    subsection fluid-fluid interaction
+      set surface tension coefficient = 0
+    end
+  end
```
Comment (AmishgaAlphonius) [inline applications_tests/lethe-fluid/cahn_hilliard_epsilon_verbosity.prm:68]:
```suggestion
  set number of fluids                = 2
  set number of material interactions = 1
```

File: doc/source/parameters/cfd/cahn_hilliard.rst [hunk 5]
```diff
@@ -38,14 +38,19 @@ The default values of the Cahn-Hilliard parameters are given in the text box bel
...
 
-* ``epsilon``: defines the :math:`\epsilon` parameter. It can either be user-defined or determined automatically for each cell. For the latter, epsilon is equal to two times the characteristic length of the cell (:math:`h`). The choices are ``automatic`` (default) or ``manual``.
+* ``epsilon``: defines the :math:`\epsilon` parameter.
+ 
+  * ``verbosity``: enables the display of the interface thickness values in the terminal. This does not affect the printing of output  fils. Choices are ``quiet`` (default, no output) or ``verbose`` (output at every iteration).
 
+  * ``method``: sets the method for the computation of epsilon. The first choice is ``automatic``: the interface thickness is determined automatically using the minimum cell diameter obtained by the deal.ii ``minimum_cell_diameter`` `method <https://www.dealii.org/current/doxygen/deal.II/namespaceGridTools.html#a47c293eff2ec7ce4b90ba08b35d1f2e2>`_. The other choice is ``manual``: allows the user to define the interface thickness by hand by setting the desired value in ``value``.
```
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/cahn_hilliard.rst:53]:
```suggestion
  * ``method``: sets the method for the computation of epsilon. The first choice is ``automatic``: the interface thickness is determined automatically using the minimum cell diameter obtained by the deal.ii ``minimum_cell_diameter`` `method <https://www.dealii.org/current/doxygen/deal.II/namespaceGridTools.html#a47c293eff2ec7ce4b90ba08b35d1f2e2>`_. The other choice is ``manual``: allows the user to define the interface thickness by hand, by setting the desired value with ``value``.
```

File: include/core/parameters_multiphysics.h [hunk 6]
```diff
@@ -63,6 +63,12 @@ namespace Parameters
     manual
   };
 
+  enum class EpsilonVerbosity
+  {
+    quiet,
+    verbose
+  };
```
Comment (AmishgaAlphonius) [inline include/core/parameters_multiphysics.h:70]: Doxygen commenting

```suggestion
  /**
   * @brief Verbosity options for the epsilon parameter.
   */
  enum class EpsilonVerbosity
  {
    /// Epsilon related information will not be displayed on terminal
    quiet,
    /// Epsilon value will be displayed on terminal for every steady and
    /// transient iteration.
    verbose
  };
```

File: source/core/parameters_multiphysics.cc [hunk 7]
```diff
@@ -519,6 +519,12 @@ Parameters::CahnHilliard::declare_parameters(ParameterHandler &prm)
         "1.0",
         Patterns::Double(),
         "Parameter linked to the interface thickness. Should always be bigger than the characteristic size of the smallest element");
+
+      prm.declare_entry(
+        "verbosity",
+        "quiet",
+        Patterns::Selection("quiet|verbose"),
+        "Display the value of epsilon for each time iteration is set to verbose");
```
Comment (AmishgaAlphonius) [inline source/core/parameters_multiphysics.cc:527]:
```suggestion
        "Display the value of epsilon for each time iteration if set to verbose");
```

File: source/solvers/cahn_hilliard.cc [hunk 8]
```diff
@@ -695,6 +695,14 @@ CahnHilliard<dim>::postprocess(bool first_iteration)
         }
     }
 
+  if (simulation_parameters.multiphysics.cahn_hilliard_parameters
+        .epsilon_verbosity == Parameters::EpsilonVerbosity::verbose)
+    {
+      double epsilon = GridTools::minimal_cell_diameter(*triangulation);
+      announce_string(this->pcout, "Epsilon value");
```
Comment (AmishgaAlphonius) [inline source/solvers/cahn_hilliard.cc:702]: Is the announce-string necessary?

**Review by AmishgaAlphonius — CHANGES_REQUESTED** [review]
Nice work! I have a few comments! :)

File: CHANGELOG.md (hunk 1 above)
Comment (PierreLaurentinCS) [reply CHANGELOG.md:10]: This was NOT intended, thanks for pointing it out

File: include/core/parameters_multiphysics.h (hunk 6 above)
Comment (PierreLaurentinCS) [reply include/core/parameters_multiphysics.h:70]: Thanks for the suggested briefs and comments

File: CHANGELOG.md [hunk 9]
```diff
@@ -9,6 +9,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/).
 
 - MINOR Remove the GIFs from the repository to reduce its size. [#1289](https://github.com/chaos-polymtl/lethe/pull/1289)
 
+### Added
```
Comment (AmishgaAlphonius) [inline CHANGELOG.md:12]:
```suggestion

## [Master] - 2024-09-25

### Added
```

File: doc/source/parameters/cfd/cahn_hilliard.rst [hunk 10]
```diff
@@ -38,14 +38,19 @@ The default values of the Cahn-Hilliard parameters are given in the text box bel
...
 
-* ``epsilon``: defines the :math:`\epsilon` parameter. It can either be user-defined or determined automatically for each cell. For the latter, epsilon is equal to two times the characteristic length of the cell (:math:`h`). The choices are ``automatic`` (default) or ``manual``.
+* ``epsilon``: defines the :math:`\epsilon` parameter.
+ 
+  * ``verbosity``: enables the display of the interface thickness values in the terminal. This does not affect the printing of output  fils. Choices are ``quiet`` (default, no output) or ``verbose`` (output at every iteration).
 
+  * ``method``: sets the method for the computation of epsilon. The first choice is ``automatic``: the interface thickness is determined automatically using the minimum cell diameter obtained by the deal.ii ``minimum_cell_diameter`` `method <https://www.dealii.org/current/doxygen/deal.II/namespaceGridTools.html#a47c293eff2ec7ce4b90ba08b35d1f2e2>`_. The other choice is ``manual``: allows the user to define the interface thickness by hand by setting the desired value with ``value``.
```
Comment (AmishgaAlphonius) [inline doc/source/parameters/cfd/cahn_hilliard.rst:53]:
```suggestion
  * ``method``: sets the method for the computation of epsilon. The first choice is ``automatic``: the interface thickness is determined automatically using the minimum cell diameter obtained by the deal.ii ``minimum_cell_diameter`` `method <https://www.dealii.org/current/doxygen/deal.II/namespaceGridTools.html#a47c293eff2ec7ce4b90ba08b35d1f2e2>`_. The other choice is ``manual``: allows the user to define the interface thickness by hand, by setting the desired value with ``value``.
```

**Review by AmishgaAlphonius — APPROVED** [review]
2 last minor comments/suggestions

File: doc/source/parameters/cfd/cahn_hilliard.rst [hunk 11]
```diff
@@ -38,14 +38,19 @@ The default values of the Cahn-Hilliard parameters are given in the text box bel
...
+      set verbosity = quiet
+      set method    = automatic
+      set value     = 1
     end
   end
   
 * ``potential smoothing coefficient``: defines the :math:`\xi` parameter in the equations above. Its value should be increased if the potential presents excessive oscillations (in advective problems for instance).
 
-* ``epsilon``: defines the :math:`\epsilon` parameter. It can either be user-defined or determined automatically for each cell. For the latter, epsilon is equal to two times the characteristic length of the cell (:math:`h`). The choices are ``automatic`` (default) or ``manual``.
+* ``epsilon``: defines the :math:`\epsilon` parameter.
```
Comment (blaisb) [inline doc/source/parameters/cfd/cahn_hilliard.rst:49]: maybe here I would add a reference to the theory guide to recall the role of the epsilon parameter. Because otherwise this is not easy to figure out

File: source/core/parameters_multiphysics.cc [hunk 12]
```diff
@@ -553,6 +559,20 @@ Parameters::CahnHilliard::parse_parameters(ParameterHandler    &prm,
...
+          CahnHilliard::epsilon_verbosity = Parameters::EpsilonVerbosity::quiet;
+        }
+      else if (op_epsilon_verbosity == "verbose")
+        {
+          CahnHilliard::epsilon_verbosity =
+            Parameters::EpsilonVerbosity::verbose;
+        }
+      else
+        throw(std::runtime_error("Invalid epsilon verbosity. "
+                                 "Options are 'quiet' or 'verbose'."));
```
Comment (blaisb) [inline source/core/parameters_multiphysics.cc:574]: Use the AssertThrow macro instead please

**Review by blaisb — CHANGES_REQUESTED** [review]
Two small comments to address

File: source/core/parameters_multiphysics.cc (hunk 12 above)
Comment (PierreLaurentinCS) [reply source/core/parameters_multiphysics.cc:574]: By looking at the code I realized we actually don't need the last part (else{ throw error}) because deal.ii already deals with the string not matching a given pattern. So we actually never enter the else. 
I think this is the same with all the other parameters that take strings as input, maybe we could clean up all of them in a subsequent PR.
If you think the "else" should be kept for a reason, please tell me

Comment (PierreLaurentinCS) [reply source/core/parameters_multiphysics.cc:574]: IRL discussion lead to leaving the exception throwing for an extra-security. Question solved

**Review by blaisb — APPROVED** [review]
